File: go/timing/timing.go
// Package timing mirrors loudkit.timing: where each chunk — and,
// approximately, each word — lands in the waveform.
//
// A reading app highlights the sentence it is speaking. That needs two
// different kinds of answer, and this package is careful to keep them apart,
// because conflating them is how a feature like this becomes a lie.
//
// Chunk times are exact. The engine renders each chunk to its own waveform and
// concatenates them, so it knows every chunk's sample offset and sample length
// without estimating anything. ChunkTiming reports those, converted to seconds.
// Chunk k's End is bit-identical to chunk k+1's Start: both are the same
// integer sample offset divided by the same sample rate, so a highlight driven
// by them can neither gap nor overlap.
//
// Word times are estimated. The model emits speech tokens, not an alignment;
// nothing in this pipeline knows where a word begins. WordTiming distributes a
// chunk's real duration across its words in proportion to how long each word is
// in code points, and that is all it is. It is right often enough to be useful
// for a highlight at sentence scale and wrong in the ways you would expect: a
// long word said fast, a short word held, a pause before a clause. The error
// grows with the length of the chunk, because a single bad guess early shifts
// everything after it — one sentence is usually fine, a long paragraph read as
// one chunk is not. If you need real alignment, you need a forced aligner; this
// is not one, and pretending otherwise would be worse than the estimate.
//
// Both are computed after any time-stretch, on the waveform the caller actually
// receives, so a result rendered at speed 1.5 needs no 1/speed correction
// applied to them.
package timing

import (
	"strings"
	"unicode/utf8"
)

// Span is what one rendered chunk contributes to a timeline.
//
// The three facts the engine has at concatenation time and nothing else: the
// text it was asked to speak (post-funnel, which is what was tokenised), how
// many samples it rendered to, and how many speech tokens it took. An input
// type rather than a per-chunk ChunkTiming, because the offsets are only
// knowable once the order is known.
type Span struct {
	Text    string
	Samples int
	Tokens  int
}

// WordTiming is one word's estimated span, in seconds from the start of the
// synthesis.
//
// Estimated, by proportional allocation. The chunk's real duration is divided
// among its words in proportion to their length in code points. There is no
// alignment model here and no per-word measurement — see the package comment
// for what that costs you.
type WordTiming struct {
	// Text is the word as it appears in the chunk, punctuation included.
	//
	// Punctuation stays attached because the split is on whitespace: a caller
	// highlighting "end." wants the full stop lit with the word, and a caller
	// matching back against their own text needs the substring to be a
	// substring.
	Text  string
	Start float64
	End   float64
}

// ChunkTiming is one chunk's exact span, and its words' estimated ones.
//
// The two tiers in one value on purpose: a caller that trusts only the exact
// tier reads Start/End and ignores Words, and the field names make it
// impossible to reach the estimate by accident.
type ChunkTiming struct {
	// Text is the chunk's text after the speech funnel — what was tokenised,
	// which is not always what the caller passed in (Polish respells embedded
	// English, and numbers are read as words).
	Text string

	// Start is seconds from the start of this synthesis's audio.
	//
	// Zero for the first chunk, and for every chunk handed to a Stream
	// callback: a streamed chunk is its own result and does not know what
	// preceded it, so the caller stitching the stream adds the offsets.
	Start float64

	End float64

	// Tokens is the speech tokens this chunk generated. Duration over tokens is
	// the pacing the postprocess detectors measure against, which is the other
	// reason to carry it.
	Tokens int

	Words []WordTiming
}

// Duration is End - Start.
func (c ChunkTiming) Duration() float64 { return c.End - c.Start }

// Shifted is this timing moved later by by seconds, words included.
//
// What a caller stitching a stream needs: each streamed chunk starts at zero,
// and the offsets are the caller's running total. Timeline does the same job in
// samples when the whole passage is in hand, which is exact; this is the best
// available when it is not.
func (c ChunkTiming) Shifted(by float64) ChunkTiming {
	words := make([]WordTiming, len(c.Words))
	for i, w := range c.Words {
		words[i] = WordTiming{Text: w.Text, Start: w.Start + by, End: w.End + by}
	}
	return ChunkTiming{
		Text:   c.Text,
		Start:  c.Start + by,
		End:    c.End + by,
		Tokens: c.Tokens,
		Words:  words,
	}
}

// Timeline lays rendered chunks end to end and times them.
//
// Offsets accumulate in samples, not seconds, and are divided by the rate once
// at the end. Accumulating seconds instead would make chunk k's End and chunk
// k+1's Start two different sums of the same floats, differing in the last bit
// — a gap or an overlap of a few nanoseconds, invisible in a test that compares
// with a tolerance and visible as a flicker in a highlight that switches on
// time >= start.
func Timeline(spans []Span, sampleRate int) []ChunkTiming {
	out := make([]ChunkTiming, 0, len(spans))
	at := 0
	for _, span := range spans {
		start := float64(at) / float64(sampleRate)
		at += span.Samples
		end := float64(at) / float64(sampleRate)
		out = append(out, ChunkTiming{
			Text:   span.Text,
			Start:  start,
			End:    end,
			Tokens: span.Tokens,
			Words:  EstimateWords(span.Text, start, end),
		})
	}
	return out
}

// EstimateWords splits text on whitespace and shares [start, end] out by length.
//
// The allocation is by code-point count, not by token count or by any acoustic
// measure: a word's characters are the only thing known here, and they
// correlate with duration well enough at sentence scale to drive a highlight.
// Code points rather than bytes so that the same text weights the same way in
// all five ports — a Polish "ł" is one character to a reader and two bytes to
// Go, and byte length would make an accented passage drift against the same
// passage read by the Python engine.
//
// Whitespace itself is not charged for — the gap between two words belongs to
// whichever side of the boundary the caller's player is on, and splitting it
// would only invent a third kind of span.
//
// Boundaries are computed from a running character total rather than by adding
// per-word durations, so the spans cannot drift: the first Start is exactly
// start, the last End is exactly end, and every interior boundary is shared by
// the two words that meet at it.
func EstimateWords(text string, start, end float64) []WordTiming {
	words := strings.Fields(text)
	total := 0
	lengths := make([]int, len(words))
	for i, w := range words {
		lengths[i] = utf8.RuneCountInString(w)
		total += lengths[i]
	}
	if total == 0 {
		return nil
	}
	span := end - start
	out := make([]WordTiming, 0, len(words))
	seen := 0
	for i, word := range words {
		at := start + span*(float64(seen)/float64(total))
		seen += lengths[i]
		out = append(out, WordTiming{
			Text:  word,
			Start: at,
			End:   start + span*(float64(seen)/float64(total)),
		})
	}
	return out
}

File: rust/tests/timing.rs
//! Timestamps: exact at the chunk, estimated at the word.
//!
//! The whole value of this feature is that a reading app can trust the first
//! tier and be told, loudly, not to trust the second in the same way. So the
//! tests are split the same way: the chunk assertions are equalities, the word
//! assertions are invariants (monotonic, inside the chunk, every word present)
//! and nothing here claims a word lands where a listener would say it does.
//!
//! Needs no assets. `loudkit::timing` is arithmetic over sample counts, which is
//! the whole of it — the engine parts that fill these in cannot run without the
//! checkpoint, the exported graphs and the onnxruntime shared library, and the
//! same is true in the other four ports.

use loudkit::timing::{estimate_words, timeline, ChunkSpan};

const SAMPLE_RATE: usize = 24_000;

fn span(text: &str, samples: usize, tokens: usize) -> ChunkSpan {
    ChunkSpan {
        text: text.to_string(),
        samples,
        tokens,
    }
}

/// A highlight that switches on `time >= start` flickers on a gap and
/// double-lights on an overlap, and both are invisible to a comparison with a
/// tolerance. Offsets accumulate as integer samples for exactly this reason.
#[test]
fn chunks_are_adjacent_to_the_last_bit() {
    let got = timeline(
        &[span("a b", 7_001, 3), span("c d e", 13_337, 5)],
        SAMPLE_RATE,
    );
    assert_eq!(got[0].start, 0.0);
    assert_eq!(got[1].start, got[0].end);
    assert_eq!(got[1].end, (7_001.0 + 13_337.0) / SAMPLE_RATE as f64);
}

#[test]
fn the_spans_cover_the_whole_render_with_nothing_left_over() {
    let got = timeline(
        &[
            span("one", 100, 1),
            span("two", 200, 2),
            span("three", 300, 3),
        ],
        SAMPLE_RATE,
    );
    let total: f64 = got.iter().map(loudkit::timing::ChunkTiming::duration).sum();
    assert!((total - 600.0 / SAMPLE_RATE as f64).abs() < 1e-12);
    assert_eq!(
        got.iter().map(|c| c.tokens).collect::<Vec<_>>(),
        vec![1, 2, 3]
    );
}

#[test]
fn an_empty_render_is_an_empty_timeline() {
    assert!(timeline(&[], SAMPLE_RATE).is_empty());
}

/// The text a chunk carries is the text that was tokenised, handed straight
/// through: the engine funnels before it splits, and the timeline reports what
/// was spoken rather than what was typed.
#[test]
fn the_span_text_reaches_the_timing() {
    let got = timeline(&[span("I have three apples.", 4_800, 7)], SAMPLE_RATE);
    assert_eq!(got[0].text, "I have three apples.");
    assert_eq!(got[0].words.len(), 4);
}

/// Every chunk of a stream starts at zero, so a caller who is stitching them
/// has to do the adding. That is the one arithmetic they should not have to
/// write themselves.
#[test]
fn shifting_moves_the_chunk_and_its_words_together() {
    let base = timeline(&[span("alpha beta", 24_000, 4)], SAMPLE_RATE).remove(0);
    let moved = base.shifted(2.5);
    assert_eq!(moved.start, base.start + 2.5);
    assert_eq!(moved.end, base.end + 2.5);
    for (was, now) in base.words.iter().zip(&moved.words) {
        assert_eq!(now.text, was.text);
        assert_eq!(now.start, was.start + 2.5);
        assert_eq!(now.end, was.end + 2.5);
    }
}

#[test]
fn words_tile_the_chunk_without_gaps() {
    let words = estimate_words("alpha beta gamma", 1.0, 4.0);
    assert_eq!(
        words.iter().map(|w| w.text.as_str()).collect::<Vec<_>>(),
        vec!["alpha", "beta", "gamma"]
    );
    assert_eq!(words[0].start, 1.0);
    assert_eq!(words[2].end, 4.0);
    for pair in words.windows(2) {
        assert_eq!(pair[0].end, pair[1].start);
    }
}

#[test]
fn times_are_monotonic_and_inside_the_chunk() {
    let words = estimate_words("a bb ccc dddd e", 2.5, 3.25);
    let mut previous = 2.5;
    for w in &words {
        assert!(2.5 <= w.start && w.start <= w.end && w.end <= 3.25);
        assert!(w.start >= previous);
        previous = w.start;
    }
}

/// The whole content of the estimate: characters stand in for seconds. Nothing
/// else here knows how long a word takes.
#[test]
fn a_longer_word_is_given_longer() {
    let words = estimate_words("hi internationalisation", 0.0, 1.0);
    assert!(words[1].end - words[1].start > words[0].end - words[0].start);
}

/// A caller highlighting `"end."` wants the full stop lit with the word, and a
/// caller matching back against their own text needs the substring to be a
/// substring.
#[test]
fn punctuation_stays_with_its_word() {
    let words = estimate_words("Hello, world!", 0.0, 1.0);
    assert_eq!(
        words.iter().map(|w| w.text.as_str()).collect::<Vec<_>>(),
        vec!["Hello,", "world!"]
    );
}

#[test]
fn no_text_is_no_words_rather_than_a_division_by_zero() {
    assert!(estimate_words("   ", 0.0, 1.0).is_empty());
    assert!(estimate_words("", 0.0, 1.0).is_empty());
}

/// The five ports count code points. A byte count would give Polish and
/// Japanese text different word weights in Rust than in Python, for text that
/// reads identically — `żółć` is four characters and eight bytes.
#[test]
fn length_is_counted_in_characters_not_bytes() {
    let words = estimate_words("aaaa żółć", 0.0, 1.0);
    let ascii = words[0].end - words[0].start;
    let accented = words[1].end - words[1].start;
    assert!((ascii - accented).abs() < 1e-12, "{ascii} vs {accented}");
}
